src/steinertree/steiner_tree.jl

```diff
@@ -0,0 +1,84 @@
+"""
+    filter_non_term_leaves!(g, term_vert)
+
+Remove edges of `g` so that all non-isolated leaves of `g` are in the set `term_vert`
+"""
+function filter_non_term_leaves!(
+    g::AbstractGraph{T},
+    term_vert::Vector{<:Integer}
+    ) where T<:Integer
+
+    is_term = falses(nv(g))
+    is_term[term_vert] .= true
+    leaves = [v for v in vertices(g) if degree(g, v) == 1 && !is_term[v]]
+
+    while !isempty(leaves)
+        leaf = pop!(leaves)
+        for v in neighbors(g, leaf)
+            rem_edge!(g, v, leaf)
+            if degree(g, v) == 1 && !is_term[v]
+                push!(leaves, v)
+            end
+        end
+    end
+
+    return g
+end
+
+"""
+	steiner_tree(g, term_vert, distmx=weights(g))
+
+Return a minimum weight spanning graph of connected, undirected graph `g` with positive edge 
+weights represented by `distmx` using [Approximate Steiner Tree](https://en.wikipedia.org/wiki/Steiner_tree_problem#Approximating_the_Steiner_tree).
+`t = length(term_vert)`
+
+### Performance
+Runtime: O(t*(t*log(t)+|E|*log(|V| ))
+Memory: O(t*|V|)
+Approximation Factor: 2-2/t
+"""
+function steiner_tree end
+
+@traitfn function steiner_tree(
+    g::AG::(!IsDirected),
+    term_vert::Vector{<:Integer},
+    distmx::AbstractMatrix{U} = weights(g)
+    ) where {U<:Real, T, AG<:AbstractGraph{T}}
+
+    nvg = nv(g)
+    term_to_actual = T.(unique(term_vert))
+
+    # Compute the graph formed by inducing the metric closure graph of g on term_vert
+    nvg_mc = length(term_to_actual)
+    distmx_mc = Matrix{U}(undef, nvg_mc, nvg_mc)
+    parents = Matrix{T}(undef, nvg, nvg_mc)
+
+    for (i, v) in enumerate(term_to_actual)
+        d_s = dijkstra_shortest_paths(g, v, distmx)
+        @inbounds parents[:, i] = d_s.parents
+        @inbounds distmx_mc[:, i] = d_s.dists[term_to_actual]
+    end
+
+    # Compute MST of Metric Closure graph
+    mst_mc = kruskal_mst(CompleteGraph(nvg_mc), distmx_mc)
```

Comment:
Is it necessary to do this on the complete graph instead of just the sub graph induced by the terminal vertices?

Reply:
Yes
For example, a subgraph induced by the terminal vertices could have no edges.

A more efficient representation of a complete graph like lazy complement of `SimpleGraph(n)` might help.

Reply:
So I don't know the algorithm, but is it the case, that distances between disconnected components in the induced subgraphs are approximated by their distance in the complete graph? Then that would work. Instead of implementing a lazy graph, it might maybe be more efficient to have an internal function that calculates the minimum spanning tree on  a complete graph.

But maybe it is also better to leave it, as it is right now and add a short TODO comment, so other people know that there might be room for improvements.